src/Documentation/resources/Best_Practices.md

```diff
@@ -0,0 +1,421 @@
+        DeleteStateOnClear (hard or soft deletions)
+
+    -   Orleans serializer efficiently stores binary data in one Azure table
+        cell
+
+    -   Data size limit == max size of the Azure column which is 64kb of binary
+        data
+
+    -   Community contributed code that extends the use of multiple table
+        columns which increases the overall maximum size to 1mb.
+
+-   SharedStorageProvider will write data to multiple underlying storage
+    providers by using the grain id hash.
+
+    -   Example: *need to create*
+
+Storage Provider Debugging Tips
+
+-   TraceOverride Verbose3 will log much more information about storage
+    operations.
+
+    -   Update silo config file
+
+        -   LogPrefix=”Storage” for all providers, or specific type using
+            “Storage.Memory” / ”Storage.Azure” / “Storage.Shard”
+
+How to deal with Storage Operation Failures
+
+-   Grains and storage providers can await storage operations and *retry*
+    failures as needed
+
+-   Unhandled failures will propagate back to the caller and will be seen by the
+    client as a broken promise
+
+-   Other than the initial read, there is not a concept that automatically
+    destroys activations if a storage operation fails
+
+-   Retrying a failing storage is *not* a default feature for built-in storage
+    providers
+
+Grain Persistence Tips
+
+Grain Size
+
+-   Optimal throughput is achieved by using *multiple smaller grains* rather
+    than a few larger grains. However, the best practice of choosing a grain
+    size and type is base on the *application domain model*.
+
+    -   Example: Users, Orders, etc.
+
+External Changing Data
+
+-   Grain are able to re-read the current state data from storage by using
+    State.ReadStateAsyc()
+
+-   A timer can also be used to re-read data from storage periodically as well
+
+    -   The functional requirements could be based on a suitable “staleness” of
+        the information
+
+        -   Example: Content Cache Grain
+
+-   Adding and Removing Fields
+
+    -   The storage provider will determine the effects of adding and removing
+        additional fields from its persisted state.
+
+    -   Azure table does not support schemas and should automatically adjust to
+        the additional fields.
+
+Writing Custom Providers
+
+-   Storage providers are simple to write which is also a significant extension
+    element for Orleans
+
+    -   Tutorial: *need tutorial*
+
+-   The API GrainState API contract drives the storage API contract (Write,
+    Clear, ReadStateAsync())
+
+-   The storage behavior is typically configurable (Batch writing, Hard or Soft
+    Deletions, etc.) and defined by the storage provider
+
+## Cluster Management
+
+-   Orleans automatically manages clusters
+
+    -   Worker roles (silo) can fail and join at any time which is handled
+        automatically
+
+    -   A silo instance table exists for diagnostics
```

Comment:
The tables exist first and foremost for the clustering protocol to work. It is also very helpful for diagnostics because it keeps history of all silos in the cluster.